### PR TITLE
Use `cli_abort()` instead of `stop()` or `abort()`

### LINKED ISSUE #837: Evaluate `stop_` calls in `map_` environments
To improve rlang backtraces. See https://github.com/r-lib/rlang/issues/1023

### LINKED ISSUE #892: Update errors to new structure

### PR DESCRIPTION
**Pull request by hadley**
Most of these changes are fairly routine, just adding snapshot tests and then calling `cli_abort()`. The `stop_bad_*()` functions, however, required more of an aggressive update to get them inline with our current standards:

* No more custom classes and custom condition data, since they were never used
* Remove `stop_bad_element_vector()` in terms of `stop_bad_element_type()` and `stop_bad_element_length()` so that we can use cli formatting. This makes the error messages slightly worse (since type and length are now reported separately) but I think the benefit of increased consistency is worth it.
* I tweaked the error message from "Element {i} of `.x` to `.x[[i]]`.

Fixes #892. Fixes #837.

### CODE REVIEW

**Comment by hadley** [comment]
There are still plenty of `Rf_errorcall()` in the C code; not sure if it's worth replacing those or not.

File: R/adverb-auto-browse.R [hunk 1]
```diff
@@ -25,7 +25,7 @@
 #'
 auto_browse <- function(.f) {
   if (is_primitive(.f)) {
-    abort("Can not auto_browse() primitive functions")
+    cli::cli_abort("Can not {.fn auto_browse} primitive functions")
```
Comment (lionel-) [inline R/adverb-auto-browse.R:28]: No need to repeat the function name.
Use "can't".
Use full stop.

```suggestion
    cli::cli_abort("Can't browse primitive functions.")
```

File: R/adverb-partial.R [hunk 2]
```diff
@@ -98,11 +98,12 @@ partial <- function(.f,
...
-    special =
-      as_closure(.f),
-    closure =
-      .f,
-    abort(sprintf("`.f` must be a function, not %s", friendly_type_of(.f)))
+    special = as_closure(.f),
+    closure = .f,
+    {
+      f <- .f
+      cli::cli_abort("{.arg .f} must be a function, not {.obj_friendly_type {f}}")
```
Comment (lionel-) [inline R/adverb-partial.R:105]: There is a typo in the class name. Should pass `arg = `.

Probably best to use `check_function()` from the rlang compat file instead.

File: R/adverb-insistently.R [hunk 3]
```diff
@@ -62,7 +62,9 @@ insistently <- function(f, rate = rate_backoff(), quiet = TRUE) {
   force(quiet)
 
   if (!is_rate(rate)) {
-    stop_bad_type(rate, "a rate", arg = "rate")
+    cli::cli_abort(
```
Comment (lionel-) [inline R/adverb-insistently.R:65]: Pass `arg = "rate"` to `abort()`.

File: R/adverb-slowly.R [hunk 4]
```diff
@@ -22,7 +22,9 @@ slowly <- function(f, rate = rate_delay(), quiet = TRUE) {
   force(quiet)
 
   if (!is_rate(rate)) {
-    stop_bad_type(rate, "a rate", arg = "rate")
+    cli::cli_abort(
```
Comment (lionel-) [inline R/adverb-slowly.R:25]: Should extract this in `check_rate()` since it's used at least two times.

File: R/as_mapper.R [hunk 5]
```diff
@@ -107,7 +107,7 @@ plucker <- function(i, default) {
   )
 }
 
-as_predicate <- function(.fn, ..., .mapper, .allow_na = FALSE) {
+as_predicate <- function(.fn, ..., .mapper, .allow_na = FALSE, .error_call = caller_env()) {
```
Comment (lionel-) [inline R/as_mapper.R:110]: Better take `.error_arg = caller_arg(.fn)` as well, and pass to `abort()`, and use in the message.

File: R/coercion.R [hunk 6]
```diff
@@ -36,7 +36,7 @@ as_vector <- function(.x, .type = NULL) {
   if (can_simplify(.x, .type)) {
     unlist(.x)
   } else {
-    stop("Cannot coerce .x to a vector", call. = FALSE)
+    cli::cli_abort("Cannot coerce .x to a vector")
```
Comment (lionel-) [inline R/coercion.R:39]:
```suggestion
    cli::cli_abort("Can't coerce {.arg .x} to a vector.")
```

File: R/conditions.R [hunk 7]
```diff
@@ -30,123 +30,27 @@ NULL
...
-    x = x,
-    expected = expected,
-    actual = actual,
-    what = what,
-    arg = arg,
-    ...,
-    .subclass = c(.subclass, "purrr_error_bad_type")
+                          call = caller_env()) {
+  what <- what %||% what_bad_object(arg)
+  cli::cli_abort(
```
Comment (lionel-) [inline R/conditions.R:37]: Here and below: Pass `arg`.

File: R/cross.R [hunk 8]
```diff
@@ -151,10 +151,9 @@ cross <- function(.l, .filter = NULL) {
     if (!is.null(.filter)) {
       is_to_filter <- do.call(".filter", unname(out[[i]]))
       if (!is_bool(is_to_filter)) {
-        abort(sprintf(
-          "The filter function must return a single logical `TRUE` or `FALSE`, not %s",
-          as_predicate_friendly_type_of(is_to_filter)
-        ))
+        cli::cli_abort(
+          "The filter function must return a single logical `TRUE` or `FALSE`, not {.obj_type_friendly {is_to_filter}}."
```
Comment (lionel-) [inline R/cross.R:155]:
```suggestion
          "The filter function must return a single `TRUE` or `FALSE`, not {.obj_type_friendly {is_to_filter}}."
```

File: R/list-modify.R [hunk 9]
```diff
@@ -64,9 +64,12 @@ list_merge <- function(.x, ...) {
   list_recurse(.x, y, c)
 }
 
-list_recurse <- function(x, y, base_f, recurse = TRUE) {
+list_recurse <- function(x, y, base_f, recurse = TRUE, error_call = caller_env()) {
   if (!is_null(names(y)) && !is_named(y)) {
-    abort("`...` arguments must be either all named, or all unnamed")
+    cli::cli_abort(
+      "`...` arguments must be either all named, or all unnamed",
```
Comment (lionel-) [inline R/list-modify.R:70]:
```suggestion
      "`...` arguments must be either all named or all unnamed.",
```

File: R/lmap.R [hunk 10]
```diff
@@ -63,12 +63,10 @@ lmap_at <- function(.x, .at, .f, ...) {
...
 }
 
-lmap_helper <- function(.x, .ind, .f, ..., .else = NULL) {
-  if (is_formula(.f)) {
-    .f <- rlang::as_function(.f)
-  }
-  if (is_formula(.else)) {
-    .else <- rlang::as_function(.else)
+lmap_helper <- function(.x, .ind, .f, ..., .else = NULL, .error_call = caller_env()) {
+  .f <- rlang::as_function(.f, call = .error_call)
```
Comment (lionel-) [inline R/lmap.R:67]: Should also pass `arg`?

File: R/lmap.R [hunk 11]
```diff
@@ -82,7 +80,10 @@ lmap_helper <- function(.x, .ind, .f, ..., .else = NULL) {
     }
 
     if (!is.list(res)) {
-      stop_bad_type(res, "list", what = paste0("Element ", i))
+      cli::cli_abort(
+        "{.code .f(.x[[{i}]])} must return a list, not {.obj_type_friendly {res}}",
```
Comment (lionel-) [inline R/lmap.R:84]:
```suggestion
        "{.code .f(.x[[{i}]])} must return a list, not {.obj_type_friendly {res}}.",
```

File: R/modify.R [hunk 12]
```diff
@@ -322,7 +322,8 @@ modify_base <- function(mapper, .x, .y, .f, ...) {
 #' @export
 modify_depth <- function(.x, .depth, .f, ..., .ragged = .depth < 0) {
   if (!is_integerish(.depth, n = 1, finite = TRUE)) {
-    abort("`.depth` must be a single number")
+    depth <- .depth
+    cli::cli_abort("{.arg .depth} must be a single number, not {.obj_type_friendly {depth}}")
```
Comment (lionel-) [inline R/modify.R:326]:
```suggestion
    cli::cli_abort("{.arg .depth} must be a single number, not {.obj_type_friendly {depth}}.", arg = ".depth")
```

File: R/pluck.R [hunk 13]
```diff
@@ -244,7 +246,7 @@ modify_in <- function(.x, .where, .f, ...) {
 assign_in <- function(x, where, value) {
   n <- length(where)
   if (n == 0) {
-    abort("`where` must contain at least one element")
+    cli::cli_abort("{.arg where} must contain at least one element")
```
Comment (lionel-) [inline R/pluck.R:249]:
```suggestion
    cli::cli_abort("{.arg where} must contain at least one element.", arg = "where")
```

File: R/rate.R [hunk 14]
```diff
@@ -200,19 +200,25 @@ signal_rate_retry <- function(rate, length, quiet) {
...
+stop_rate_excess <- function(rate, error_call = caller_env()) {
   i <- rate_count(rate)
 
   # Bump counter to get an expired error next time around
   rate_bump_count(rate)
 
-  msg <- sprintf("Request failed after %d attempts", i)
-  abort(msg, "purrr_error_rate_excess", rate = rate)
+  cli::cli_abort(
+    "Request failed after {i} attempts",
```
Comment (lionel-) [inline R/rate.R:220]:
```suggestion
    "Request failed after {i} attempts.",
```

File: R/reduce.R [hunk 15]
```diff
@@ -200,12 +200,15 @@ reduce_early <- function(out, prev, acc, acc_out, acc_idx, left = TRUE) {
...
 
-reduce_init <- function(x, init, left = TRUE) {
+reduce_init <- function(x, init, left = TRUE, error_call = caller_env()) {
   if (!missing(init)) {
     init
   } else {
     if (is_empty(x)) {
-      stop("`.x` is empty, and no `.init` supplied", call. = FALSE)
+      cli::cli_abort(
+        "{.arg .x} is empty, and no {.arg .init} supplied.",
```
Comment (lionel-) [inline R/reduce.R:209]:
```suggestion
        "Must supply {.arg .init} when {.arg .x} is empty.",
```

File: src/flatten.c [hunk 16]
```diff
@@ -11,7 +11,7 @@ const char* objtype(SEXP x) {
 
 SEXP flatten_impl(SEXP x) {
   if (TYPEOF(x) != VECSXP) {
-    stop_bad_type(x, "a list", NULL, ".x");
+    stop_bad_type(x, "a list", NULL, ".x", R_BaseEnv);
```
Comment (lionel-) [inline src/flatten.c:14]: Could vendor the rlang C library to call `r_peek_frame()`. Or reimplement it with the base lib:

```c
r_obj* r_peek_frame() {
  return r_eval(peek_frame_call, r_envs.empty);
}

  // Initialisation
  r_obj* current_frame_body = KEEP(r_parse_eval("as.call(list(sys.frame, -1))", r_envs.base));
  r_obj* current_frame_fn = KEEP(r_new_function(r_null, current_frame_body, r_envs.empty));
  peek_frame_call = r_call(current_frame_fn);
  r_preserve(peek_frame_call);
```

Comment (lionel-) [reply src/flatten.c:14]: In general I don't pass frames around and just peek the frame at throw site. This is what `r_abort()` does. This assumes that `.Call()` was called in the relevant user facing function.

If instead we need to forward a call, then I pass a `struct r_lazy` where `.x` either contains a frame environment (`.env` is NULL or the empty env, this represents a forced promise): https://github.com/r-lib/vctrs/blob/7c9d973b3c432da00329ae9a1873236cabd07abb/src/subscript-loc.c#L435

Or a symbol to be evaluated in a frame and that resolves to a `call` argument: https://github.com/r-lib/vctrs/blob/7c9d973b3c432da00329ae9a1873236cabd07abb/src/subscript-loc.c#L450

**Comment by lionel-** [comment]
General comment: Should pass `arg = ` to `abort()` so that rlang can highlight the failing arguments in backtraces. Cf this example where `baz` is highlighted in magenta:
Screenshot 2022-09-12 at 09 20 26

File: R/lmap.R (hunk 10 above)
Comment (hadley) [reply R/lmap.R:67]: It's always `.f`?

File: src/flatten.c (hunk 16 above)
Comment (hadley) [reply src/flatten.c:14]: How about I merge this PR (since it's massive) and then tackle the C stuff separately?

File: R/lmap.R (hunk 10 above)
Comment (lionel-) [reply R/lmap.R:67]: I meant for resilience to refactoring / new usages, but that's fine as it is.

**Comment by lionel-** [comment]
I still see a lot of missing `arg = ` for `abort()`. Are you waiting to see if `cli_abort()` may handle these arguments automatically?

**Comment by hadley** [comment]
@lionel- just did a pass to add them consistently. I'll merge this and open an issue to reconsider the errors in C.

**Pull request merged**